dxr/vcs.py

```diff
@@ -327,8 +329,123 @@ def get_contents(cls, path, revision, stderr=None):
                                         filename + '@' + revision],
                                        cwd=directory, env=env, stderr=stderr)
 
+class Subversion(Vcs):
+    command = 'svn'
+
+    def __init__(self, root):
+        super(Subversion, self).__init__(root)
+        # All tracked files as a dictionary (file => revision).
+        self.files = self.__get_files_with_revision()
+        # Determine repository revision once.
+        self.revision = self.__get_revision()
+        # Determine upstream URL once.
+        self.upstream = self.__get_upstream_url()
+
+    # Interface class methods.
+
+    @classmethod
+    def claim_vcs_source(cls, path, dirs, tree):
+        # Requires ".svn" directory.
+        if '.svn' not in dirs:
+            return None
+
+        # Requires successful "svn info".
+        try:
+            cls.invoke_vcs(['info'], path)
+        except subprocess.CalledProcessError:
+            return None
+
+        # Exclude ".svn" from indexing.
+        dirs.remove('.svn')
+
+        return cls(path)
+
+    @classmethod
+    def get_contents(cls, path, revision, stderr=None):
+        directory, filename = split(path)
+        args = ['cat', '-r', str(revision), os.path.abspath(path)]
+        return cls.invoke_vcs(args, directory)
+
+    # Interface instance methods.
+
+    def is_tracked(self, path):
+        return path in self.files
+
+    def has_upstream(self):
+        return self.upstream != ""
```

Comment:
It looks like on error the upstream constructor could return None, rather than "".